app/knight.py

```diff
@@ -0,0 +1,33 @@
+from __future__ import annotations
+
+
+class Knight:
+    def __init__(self, knight: dict) -> None:
+        self.knight = knight
+
+    def preparation(self) -> None:
+        self.name = self.knight["name"]
+        self.power = (self.knight["power"]
+                      + (self.knight["weapon"].get("power", 0)
+                         if self.knight["weapon"] else 0)
+                      + (self.knight["potion"]["effect"].get("power", 0)
+                         if self.knight["potion"] else 0))
+        self.hp = (self.knight["hp"]
+                   + (self.knight["potion"]["effect"].get("hp", 0)
+                      if self.knight["potion"] else 0))
+        self.protection = ((sum(protec["protection"]
+                                for protec in self.knight["armour"])
+                            if self.knight["armour"] else 0)
+                           + (self.knight["potion"]["effect"].get("protection"
+                                                                  , 0)
+                              if self.knight["potion"] else 0))
+
+    def reduce_power(self, other: Knight) -> Knight:
+        knight = Knight(self.knight)
+        knight.power = self.power - other.protection
+        return knight
+
+    def fight(self, other: Knight) -> Knight:
+        self.hp -= other.power
+        self.hp = max(self.hp, 0)

```

Comment:
The `fight` method mutates the `Knight` instance. It's a good practice to avoid mutation where possible. Consider returning a new instance with the updated hp instead of mutating the existing one. This is a code quality issue.